design/worker-config-drive.md

```diff
@@ -0,0 +1,178 @@
+         * [User Stories [optional]](#user-stories-optional)
+            * [Story 1](#story-1)
+            * [Story 2](#story-2)
+         * [Implementation Details/Notes/Constraints [optional]](#implementation-detailsnotesconstraints-optional)
+         * [Risks and Mitigations](#risks-and-mitigations)
+      * [Design Details](#design-details)
+         * [Work Items](#work-items)
+         * [Dependencies](#dependencies)
+         * [Test Plan](#test-plan)
+         * [Upgrade / Downgrade Strategy](#upgrade--downgrade-strategy)
+         * [Version Skew Strategy](#version-skew-strategy)
+      * [Drawbacks [optional]](#drawbacks-optional)
+      * [Alternatives [optional]](#alternatives-optional)
+      * [References](#references)
+
+<!-- Added by: stack, at: 2019-02-15T11:41-05:00 -->
+
+<!--te-->
+
+## Summary
+
+In order to bring a host into the cluster as a worker we have to give
+it the correct ignition configuration as it boots. The contents of the
+Ignition file are available via the kubernetes API. Something needs to
+build a config drive image containing the ignition file and the
+baremetal operator needs to pass that formatted image to Ironic as
+part of provisioning the host.
+
+In order to avoid having the abstraction around Ironic break down,
+something that knows about the expectations of the image being
+deployed (especially the path where the Ignition file should be
+written within the config drive image) should build the drive image,
+and pass it to the baremetal operator via the host along with the
+other provisioning instructions like the image name and checksum.
+
+The Ignition file contents are stored in a Secret within the
+kubernetes database because they contain certificate authority
+information. We will need to store the config drive contents in a
+secret as well, to avoid leaking the data when we pass the drive image
+to the baremetal operator.
+
+## Motivation
+
+### Goals
+
+- Avoid having the baremetal operator tightly coupled to any specific
+  image.
+- Avoid having the baremetal operator tightly coupled to provisioning
+  hosts to become nodes in the cluster.
+- Avoid leaking secrets when passing the config drive to the baremetal
+  operator.
+
+### Non-Goals
+
+N/A
+
+## Proposal
+
+### Implementation Details/Notes/Constraints
+
+Under OpenShift, the worker settings come from the contents of a
+secret (`openshift-machine-api/worker-user-data`), which holds a
+base64-encoded copy of the Ignition JSON data.
+
+Before allocating a host to be a worker, the actuator will identify
+the Secret containing the worker settings, access contents, and create
+an ISO image with the data in the right layout. It will then read the
+ISO image into memory and compress it, before writing it to a new
+Secret. It will include a reference to that second Secret with the
+other provisioning instructions (image name, checksum, etc.) it gives
+to the baremetal operator when it allocates a host. The baremetal
```

Comment:
Shouldn't this all be done on the fly because the CA data rotates every so many hours?

Reply:
Ah, good question. Perhaps it should? That would make managing the lifecycle of the secret easier, since it could be tied to the Host object.